adaptive/learner/balancing_learner.py

```diff
@@ -112,25 +112,33 @@ def strategy(self, strategy):
                 ' strategy="npoints" is implemented.')
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        chosen_points = []
+        chosen_loss_improvements = []
+        npoints_per_learner = defaultdict(int)
+
         for _ in range(n):
             improvements_per_learner = []
-            pairs = []
+            points_per_learner = []
             for index, learner in enumerate(self.learners):
                 if index not in self._points:
                     self._points[index] = learner.ask(
                         n=1, tell_pending=False)
-                point, loss_improvement = self._points[index]
-                improvements_per_learner.append(loss_improvement[0])
-                pairs.append((index, point[0]))
-            x, l = max(zip(pairs, improvements_per_learner),
-                       key=itemgetter(1))
-            points.append(x)
-            loss_improvements.append(l)
-            self.tell_pending(x)
-
-        return points, loss_improvements
+                points, loss_improvements = self._points[index]
+                npoints = npoints_per_learner[index] + learner.npoints
+                sort_tuple = (loss_improvements[0], -npoints)
+                improvements_per_learner.append(sort_tuple)
+                points_per_learner.append((index, points[0]))
+
+            # Chose the optimal improvement.
+            (index, point), (loss_improvement, _) = max(
+                zip(points_per_learner, improvements_per_learner),
+                key=itemgetter(1))
+            npoints_per_learner[index] += 1
```

Comment:
Does this ever get decremented?

Reply:
No it won't.

Reply:
No it won't.

Reply:
But then what does it mean? Above I see `npoints_per_learner[index] + learner.npoints`

Reply:
learner.npoints doesn't change when we add a point to chosen_points, so without this all suggested points would come from the same learner.

Reply:
learner.npoints doesn't change when we add a point to chosen_points, so without this all suggested points would come from the same learner.

Reply:
But shouldn't we decrement this when a point is added to the learner? Also what is the meaning of this variable? It's not at all clear from how it's used, and seems to duplicate `learner.npoints`.

Reply:
but `learner.npoints` doesn't increase when adding something to `tell_pending`.

Reply:
So in this loop `for index, learner in enumerate(self.learners):` we ask a point from each learner. After this we can pick the `learner` and `point` with the highest `loss_improvement`, this means that in our `chosen_points` list we will have one point for the `learner` we chose. However, `learner.npoints` wouldn't have increased at this point in time! So the next time we execute this loop we need to take the `chosen_points` into account.

Reply:
Ah! I overlooked that this isn't an attribute.